test/functional/services/combo_box.js

```diff
@@ -0,0 +1,125 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+export function ComboBoxProvider({ getService }) {
+  const testSubjects = getService('testSubjects');
+  const find = getService('find');
+  const log = getService('log');
+  const retry = getService('retry');
+
+  // wrapper around EuiComboBox interactions
+  class ComboBox {
+
+    async set(comboBoxSelector, value) {
+      log.debug(`comboBox.set, comboBoxSelector: ${comboBoxSelector}`);
+      const comboBox = await testSubjects.find(comboBoxSelector);
+      await this.setElement(comboBox, value);
+    }
+
+    async setElement(comboBoxElement, value) {
+      log.debug(`comboBox.setElement, value: ${value}`);
+      await this._filterOptionsList(comboBoxElement, value);
+      await find.clickByCssSelector('.euiComboBoxOption');
+      await this.closeOptionsList(comboBoxElement);
+    }
+
+    async filterOptionsList(comboBoxSelector, filterValue) {
+      log.debug(`comboBox.filterOptionsList, comboBoxSelector: ${comboBoxSelector}, filter: ${filterValue}`);
+      const comboBox = await testSubjects.find(comboBoxSelector);
+      await this._filterOptionsList(comboBox, filterValue);
+      await this.closeOptionsList(comboBox);
+    }
+
+    async _filterOptionsList(comboBoxElement, filterValue) {
+      const input = await comboBoxElement.findByTagName('input');
+      await input.clearValue();
+      await input.type(filterValue);
+      await comboBoxElement.waitForDeletedByClassName('euiLoadingSpinner');
```

Comment:
I understand here where after typing a filterValue you are waiting for the euiLoadingSpinner.  But what I'm wondering is this service should be checking that loading indicator in the beginning of some methods so that we know the initial populating of the list has completed.  Maybe in `doesComboBoxHaveSelectedOptions`?  Or in the beginning of the  `set` or `setElement` methods?

Reply:
`getOptionsList` should also wait for the options list to load,  nice catch.

`doesComboBoxHaveSelectedOptions` will not need to wait for the options list to load because it does not use the options list. The selected options are not loaded async.

`set` and `setElement` methods will not need to wait for the options list because they use the `_filterOptionsList` method which will handle this for them.

`_filterOptionsList` will not need to wait at the beginning because combo box input can be used while options are being fetched.